docs/mapping/cawe/dialogs/options.rst
.. _the_configure_cawe_options_dialog:

The Configure CaWE Options Dialog
=================================

This dialog is opened by selecting **Configure CaWE…** from
:ref:`The File Menu <mapping_cawe_menureference_file_the_file_menu>`.

It is used to configure the general appearance and handling of CaWE.

Game Configurations
-------------------

|image0|

This tab contains configuration options for game configurations. Each
map has to use a game configuration in order to work.

Game configurations are automatically created from the subdirectories of
/Games that have a game data file (.fgd) with the same name as the
directory in it. For example /Games/Deathmatch/Deathmatch.fgd is a valid
game configuration with the name Deathmatch.

In this tab you can edit properties for game configurations that will be
used when editing a map that uses this game configuration.

Game Configuration Selection
   Selects the game configuration to adjust its properties.

Default Point Entity class
   Sets the Point Entity class that is selected by default when a new
   map is opened/created.
Default Brush Entity class
   Sets the Brush Entity class that is selected by default when a new
   map is opened/created.
Default texture scale
   Sets the default texture scale for this profile.
Default lightmap scale
   Sets the default lightmap scale for this profile.
Cordon texture
   This option is not implemented at this time.

General
-------

|image1|

This tab contains general configuration options for the editor to
control its behavior. The paths to the executables in this tab are very
important so CaWE can find them to compile and run edited maps.

Undo Levels
   Sets the maximum number of actions performed by the editor that can
   be revoked.

Allow grouping/ungrouping while Ignore Groups is checked
   Toggles if you can group or ungroup selected objects even if the
   **Ignore Groups** Option from the
   :ref:`Selection Tool <the_selection_tool>` options bar is checked.

Stretch arches to fit original bounding rectangle
   This option stretches a arch to fit the bounding rectangle when
   created. If this options is not selected the size of the arch will be
   determined by the size of the bounding rectangle as well as the
   parameters set in the arch creation dialog (see
   :ref:`The New Brush Tool <the_new_brush_tool>`).

Executables
~~~~~~~~~~~

The path of executables used to compile and run maps can be configured
here in case the user changed their paths or wants to use different
ones.

Engine executable
   Sets the path to the Cafu executable.
CaBSP executable
   Sets the path to the CaBSP executable.
CaPVS executable
   Sets the path to the CaPVS executable.
CaLight executable
   Sets the path to the CaLight executable.

At the bottom of this tab, the path to your configuration data is shown.

.. _d_views:

2D Views
--------

|image2|

This tab holds configuration options that define the appearance of the
editors 2D views.

Options
~~~~~~~

Crosshair cursor
   This option is not implemented at this time.
Default to 15 degree rotations
   If rotating an object this sets the rotation steps to 15 degree
   instead of smooth rotating. Note that you can still rotate objects
   smoothly by pressing ``SHIFT`` while rotating.
Display scrollbars
   Toggles the display of scrollbars in the 2D views.
Draw vertices
   Toggles display of objects vertices in a 2D view.
White-on-Black color scheme
   Switches between white grid on black background and black grid on
   white background modes.
Keep group when done dragging
   When cloning objects from a VisGroup, this option adds the cloned
   objects to the same VisGroup after dragging them.
Center on camera after movement in 3D
   This option is not implemented at this time.
Use Visgroup colors for object lines
   Toggles if the line color defined in the Edit VisGroups dialog is
   used to display the objects that belong to this group in a 2D view.
Arrow keys nudge selected object/vertice
   If activated the arrow keys can be used to move objects or vertices.
Reorient primitives on creation in the active 2D view
   When creating a brush in a 2D view, this brushes top side is per
   default oriented in the direction of the top down 2D view. If this
   option is activated the top side of the brush will be oriented in the
   direction of the 2D view in which the brush has been created.
Automatic infinite selection in 2D windows (no ENTER)
   If dragging a selection box, this option determines if the objects in
   the selection box are selected instantly after the mouse button is
   released.
Selection box selects by center handles only
   Determines if the center handle of objects has to lie within a
   selection box for the object to be selected. Otherwise any object
   that lies partially in the selection box is selected.

Grid
~~~~

Size
   Defines the default grid size.
Intensity
   Adjusts the grids color intensity.
Highlight every 64 units
   If checked, a grid line is highlighted every 64 units.
Highlight every 1024 units
   If checked, a grid line is highlighted every 1024 units.
Highlight every *X* grid lines
   If a grid line is the *X*. grid line in a row, it is highlighted.
Hide grid smaller than 4 pixel
   If the distance between two grid lines gets smaller than 4 pixels in
   a 2D view, the grid lines are no longer displayed.
Dotted Grid
   Toggles between a grid with solid lines or dots.

.. _d_views1:

3D Views
--------

|image3|

This tab holds configuration options that define the appearance of the
editors 3D view.

Performance
~~~~~~~~~~~

Animate models
   With this option activated, models are animated in the editor.
Back clipping plane
   Sets the back clipping plane in the editors 3D view. With high values
   even geometry far away from the camera is rendered. This option has a
   huge impact on the performance of the editor, especially in big maps.
Model render distance
   Determines the distance at which models are rendered as bounding
   boxes instead of the real model meshes.

Navigation
~~~~~~~~~~

Reverse mouse Y axis (aircraft style)
   Toggles reversion of the Y axis of the mouse, when changing the
   cameras view direction in a 3D view with the mouse.
Forward speed
   Sets the maximum speed when moving forward in a 3D view.
Time to top speed (msec)
   Adjusts the time until a movement in a 3D view speeds up to its
   maximum speed.

.. |image0| image:: /images/mapping/cawe/dialogs/dialog_configure_gameconfig.png
   :class: medialeft
.. |image1| image:: /images/mapping/cawe/dialogs/dialog_configure_general.png
   :class: medialeft
.. |image2| image:: /images/mapping/cawe/dialogs/dialog_configure_2dviews.png
   :class: medialeft
.. |image3| image:: /images/mapping/cawe/dialogs/dialog_configure_3dviews.png
   :class: medialeft



CaWE/Options.cpp
/*
Cafu Engine, http://www.cafu.de/
Copyright (c) Carsten Fuchs and other contributors.
This project is licensed under the terms of the MIT license.
*/

#include "Options.hpp"
#include "GameConfig.hpp"

#include "wx/wx.h"
#include "wx/confbase.h"
#include "wx/datetime.h"
#include "wx/fileconf.h"
#include "wx/stdpaths.h"
#include "wx/dir.h"
#include "wx/filename.h"


OptionsT::~OptionsT()
{
    DeleteGameConfigs();
}


static bool CompareGameConfigs(GameConfigT* const& GC1, GameConfigT* const& GC2)
{
    return wxStricmp(GC1->Name, GC2->Name)<0;
}


static wxString FindTool(const wxString& Tool, const wxString& Fallback)
{
#ifdef __WXMSW__
    const char* Compilers[] = { "vc14", "vc12", "vc11", "vc10", "vc9", "vc8", NULL };
    const char* Archs[]     = { "x64", "x86", NULL };

    for (const char** Compiler = Compilers; *Compiler; Compiler++)
        for (const char** Arch = Archs; *Arch; Arch++)
        {
            const wxString Path = wxString("build/win32/") + (*Compiler) + "/" + (*Arch) + "/release/" + Tool;

            if (wxFileExists(Path)) return Path;
        }
#else
    const char* Compilers[] = { "g++", NULL };

    for (const char** Compiler = Compilers; *Compiler; Compiler++)
    {
        const wxString Path = wxString("build/linux2/") + (*Compiler) + "/release/" + Tool;

        if (wxFileExists(Path)) return Path;
    }
#endif

    return Fallback;
}


void OptionsT::Init()
{
    time_t Ticks=wxConfigBase::Get()->Read("Configured/Installed", 0L);

    if (Ticks==0)
    {
        // If we are reading this config for the very first time,
        // set the "ticks since installed" to the current time.
        Ticks=wxDateTime::Now().GetTicks();
        wxConfigBase::Get()->Write("Configured/Installed", long(Ticks));
    }

    DaysSinceInstalled=wxDateTime::Now().Subtract(wxDateTime(Ticks)).GetDays();


    // Read settings.
    wxConfigBase::Get()->SetPath("General");
    wxConfigBase::Get()->Read("Undo Levels",       &general.UndoLevels,        50   );
    wxConfigBase::Get()->Read("Locking Textures",  &general.LockingTextures,   true );
    wxConfigBase::Get()->Read("UVs Face Aligned",  &general.NewUVsFaceAligned, false);
#ifdef __WXMSW__
    wxConfigBase::Get()->Read("Engine Executable", &general.EngineExe, FindTool("Ca3DE/Cafu.exe",      "Cafu.exe"));
    wxConfigBase::Get()->Read("BSP Executable",    &general.BSPExe,    FindTool("CaBSP/CaBSP.exe",     "CaBSP.exe"));
    wxConfigBase::Get()->Read("Light Executable",  &general.LightExe,  FindTool("CaLight/CaLight.exe", "CaLight.exe"));
    wxConfigBase::Get()->Read("PVS Executable",    &general.PVSExe,    FindTool("CaPVS/CaPVS.exe",     "CaPVS.exe"));
#else
    wxConfigBase::Get()->Read("Engine Executable", &general.EngineExe, FindTool("Ca3DE/Cafu",      "./Cafu"));
    wxConfigBase::Get()->Read("BSP Executable",    &general.BSPExe,    FindTool("CaBSP/CaBSP",     "./CaBSP"));
    wxConfigBase::Get()->Read("Light Executable",  &general.LightExe,  FindTool("CaLight/CaLight", "./CaLight"));
    wxConfigBase::Get()->Read("PVS Executable",    &general.PVSExe,    FindTool("CaPVS/CaPVS",     "./CaPVS"));
#endif
    wxConfigBase::Get()->SetPath("..");

    wxConfigBase::Get()->SetPath("Views2D");
    wxConfigBase::Get()->Read("Draw Vertices",     &view2d.DrawVertices,      true );
    wxConfigBase::Get()->Read("SelectByHandles",   &view2d.SelectByHandles,   false);
    wxConfigBase::Get()->Read("ShowEntityInfo",    &view2d.ShowEntityInfo,    true );
    wxConfigBase::Get()->Read("ShowEntityTargets", &view2d.ShowEntityTargets, false);
    wxConfigBase::Get()->Read("UseGroupColors",    &view2d.UseGroupColors,    true );
    wxConfigBase::Get()->SetPath("..");

    wxConfigBase::Get()->SetPath("Views3D");
    wxConfigBase::Get()->Read("Reverse Y",         &view3d.ReverseY,          false);
    wxConfigBase::Get()->Read("BackPlane",         &view3d.BackPlane,         6000);
    wxConfigBase::Get()->Read("ModelDistance",     &view3d.ModelDistance,     3000);
    wxConfigBase::Get()->Read("AnimateModels",     &view3d.AnimateModels,     true);
    wxConfigBase::Get()->Read("MaxCameraVelocity", &view3d.MaxCameraVelocity, 1000);
    wxConfigBase::Get()->Read("TimeToMaxSpeed",    &view3d.TimeToMaxSpeed,    500);
    wxConfigBase::Get()->Read("MouseSensitivity",  &view3d.MouseSensitivity,  0.4f);
    wxConfigBase::Get()->Read("SplitPlanesDepth",  &view3d.SplitPlanesDepth,  0);
    wxConfigBase::Get()->SetPath("..");

    wxConfigBase::Get()->SetPath("Grid");
    wxConfigBase::Get()->Read("InitialSpacing",    &Grid.InitialSpacing,    8);
    wxConfigBase::Get()->Read("MinPixelSpacing",   &Grid.MinPixelSpacing,   4);
    wxConfigBase::Get()->Read("UseDottedGrid",     &Grid.UseDottedGrid,     false);
    wxConfigBase::Get()->Read("ShowHighlight1",    &Grid.ShowHighlight1,    true);
    wxConfigBase::Get()->Read("SpacingHighlight1", &Grid.SpacingHighlight1, 64);
    wxConfigBase::Get()->Read("ShowHighlight2",    &Grid.ShowHighlight2,    true);
    wxConfigBase::Get()->Read("SpacingHighlight2", &Grid.SpacingHighlight2, 1024);
    wxConfigBase::Get()->Read("ColorBackground",   &Grid.ColorBackground,   wxColour(0, 0, 0));
    wxConfigBase::Get()->Read("ColorBaseGrid",     &Grid.ColorBaseGrid,     wxColour(32, 32, 32));
    wxConfigBase::Get()->Read("ColorHighlight1",   &Grid.ColorHighlight1,   wxColour(64, 64, 64));
    wxConfigBase::Get()->Read("ColorHighlight2",   &Grid.ColorHighlight2,   wxColour(90, 53, 37));
    wxConfigBase::Get()->Read("ColorAxes",         &Grid.ColorAxes,         wxColour(37, 90, 37));
    wxConfigBase::Get()->SetPath("..");

    // Assign the remaining colors.
    colors.Vertex       =wxColour(255, 255, 255);
    colors.ToolHandle   =wxColour(255, 255, 255);
    colors.ToolDrag     =wxColour(255, 255, 0);
    colors.ToolSelection=wxColour(255, 0, 0);
    colors.Selection    =wxColour(255, 0, 0);
    colors.SelectedFace =wxColour(220, 0, 0);
    colors.SelectedEdge =wxColour(250, 220, 0);
    colors.ToolMorph    =wxColour(255, 0, 0);


    wxFileConfig CfgFile(wxTheApp->GetAppName(), wxTheApp->GetVendorName(), wxStandardPaths::Get().GetUserDataDir()+"/Games.config");
    wxDir        GamesDir("Games");     // Should also work under Linux, otherwise use "./Games".
    wxString     GameName;

    // Iterate through all subdirectories of "Games/" and create a game configuration for each.
    // The creation of a game configuration succeeds (that is, no exception is thrown) only when
    // the related "EntityClassDefs.lua" script could properly be loaded and processed.
    // Other game configuration specific data is initialized from the relevant section from the CfgFile.
    for (bool more=GamesDir.GetFirst(&GameName, "", wxDIR_DIRS); more; more=GamesDir.GetNext(&GameName))
    {
        CfgFile.SetPath("GameConfig_"+GameName);

        try
        {
            GameConfigs.PushBack(new GameConfigT(CfgFile, GameName, "Games/"+GameName));
        }
        catch (const GameConfigT::InitErrorT&)
        {
            // Just do nothing but warn the user if there was an error initializing the game config,
            // that is, the game config will not be pushed back to the GameConfigs array.
            wxMessageBox("Configuration for game \""+GameName+"\" could not be loaded!", "Warning", wxOK | wxICON_EXCLAMATION);
        }

        CfgFile.SetPath("..");
    }

    GameConfigs.QuickSort(CompareGameConfigs);

    // If we couldn't load any game configurations, pop up the options screen.
    /* if (GameConfigs.Size()==0)
    {
        wxMessageBox("I scanned the Games subdirectory for game configurations,\n"
                     "but didn't find any.\n"
                     "You need to have at least one game configuration inside your\n"
                     "Games directory for the editor to work.\n"
                     "Please re-install Cafu or contact the Cafu forums for help.", "No game configurations found", wxOK | wxICON_ERROR);
    } */
}


void OptionsT::Write() const
{
    wxConfigBase::Get()->SetPath("General");
    wxConfigBase::Get()->Write("Undo Levels",       general.UndoLevels       );
    wxConfigBase::Get()->Write("Locking Textures",  general.LockingTextures  );
    wxConfigBase::Get()->Write("UVs Face Aligned",  general.NewUVsFaceAligned);
    wxConfigBase::Get()->Write("Engine Executable", general.EngineExe        );
    wxConfigBase::Get()->Write("BSP Executable",    general.BSPExe           );
    wxConfigBase::Get()->Write("Light Executable",  general.LightExe         );
    wxConfigBase::Get()->Write("PVS Executable",    general.PVSExe           );
    wxConfigBase::Get()->SetPath("..");

    wxConfigBase::Get()->SetPath("Views2D");
    wxConfigBase::Get()->Write("Draw Vertices",     view2d.DrawVertices     );
    wxConfigBase::Get()->Write("SelectByHandles",   view2d.SelectByHandles  );
    wxConfigBase::Get()->Write("ShowEntityInfo",    view2d.ShowEntityInfo   );
    wxConfigBase::Get()->Write("ShowEntityTargets", view2d.ShowEntityTargets);
    wxConfigBase::Get()->Write("UseGroupColors",    view2d.UseGroupColors   );
    wxConfigBase::Get()->SetPath("..");

    wxConfigBase::Get()->SetPath("Views3D");
    wxConfigBase::Get()->Write("Reverse Y",         view3d.ReverseY         );
    wxConfigBase::Get()->Write("BackPlane",         view3d.BackPlane        );
    wxConfigBase::Get()->Write("ModelDistance",     view3d.ModelDistance    );
    wxConfigBase::Get()->Write("AnimateModels",     view3d.AnimateModels    );
    wxConfigBase::Get()->Write("MaxCameraVelocity", view3d.MaxCameraVelocity);
    wxConfigBase::Get()->Write("TimeToMaxSpeed",    view3d.TimeToMaxSpeed   );
    wxConfigBase::Get()->Write("MouseSensitivity",  view3d.MouseSensitivity );
    wxConfigBase::Get()->Write("SplitPlanesDepth",  view3d.SplitPlanesDepth );
    wxConfigBase::Get()->SetPath("..");

    wxConfigBase::Get()->SetPath("Grid");
    wxConfigBase::Get()->Write("InitialSpacing",    Grid.InitialSpacing     );
    wxConfigBase::Get()->Write("MinPixelSpacing",   Grid.MinPixelSpacing    );
    wxConfigBase::Get()->Write("UseDottedGrid",     Grid.UseDottedGrid      );
    wxConfigBase::Get()->Write("ShowHighlight1",    Grid.ShowHighlight1     );
    wxConfigBase::Get()->Write("SpacingHighlight1", Grid.SpacingHighlight1  );
    wxConfigBase::Get()->Write("ShowHighlight2",    Grid.ShowHighlight2     );
    wxConfigBase::Get()->Write("SpacingHighlight2", Grid.SpacingHighlight2  );
    wxConfigBase::Get()->Write("ColorBackground",   Grid.ColorBackground    );
    wxConfigBase::Get()->Write("ColorBaseGrid",     Grid.ColorBaseGrid      );
    wxConfigBase::Get()->Write("ColorHighlight1",   Grid.ColorHighlight1    );
    wxConfigBase::Get()->Write("ColorHighlight2",   Grid.ColorHighlight2    );
    wxConfigBase::Get()->Write("ColorAxes",         Grid.ColorAxes          );
    wxConfigBase::Get()->SetPath("..");

    // Write the game configs. This is in an external file so we can distribute it with CaWE as a set of defaults.
    wxFileConfig CfgFile(wxTheApp->GetAppName(), wxTheApp->GetVendorName(), wxStandardPaths::Get().GetUserDataDir()+"/Games.config");

    for (unsigned long ConfigNr=0; ConfigNr<GameConfigs.Size(); ConfigNr++)
    {
        // Each game configuration is stored in an own group (directory).
        CfgFile.SetPath("GameConfig_"+GameConfigs[ConfigNr]->Name);
        GameConfigs[ConfigNr]->Save(CfgFile);
        CfgFile.SetPath("..");
    }
}


void OptionsT::DeleteGameConfigs()
{
    for (unsigned long i=0; i<GameConfigs.Size(); i++)
        delete GameConfigs[i];

    GameConfigs.Clear();
}
